File: data-structures/stacks/visual-basic/stack_linked_list_OOP.vb
' Raspberry Pi Foundation
' Usage licensed under CC BY-SA 4
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


' A class to represent a node in a linked list
Class Node
    Private data As String
    Private nextItem As Node

    ' Constructor method
    Public Sub New(ByVal itemData As String)
        data = itemData
    End Sub


    Public Function GetData() As String
        Return data
    End Function


    Public Function GetNext() As Node
        Return nextItem
    End Function


    Public Sub SetNext(ByVal newNext As Node)
        nextItem = newNext
    End Sub


    ' Code just for testing
    Public Sub OutputNode()
        Dim nextData = Nothing
        If nextItem IsNot Nothing Then
            nextData = nextItem.data
        End If
        Console.WriteLine($"Data: {data}, next: {nextData}")
    End Sub
End Class


' A class to represent a stack implemented as a linked list
Class Stack
    Private top As Node

    ' Check if the stack is empty
    Public Function IsEmpty() As Boolean
        If top Is Nothing Then
            Return True
        Else
            Return False
        End If
    End Function

    ' Push a new item onto the stack
    Public Sub Push(ByVal data As String)
        Dim newNode As Node = New Node(data)

        ' Check if the stack is NOT empty
        If Not IsEmpty() Then
            ' Adjust the pointer to the next node
            newNode.SetNext(top)
        End If

        ' Set top to point to the new node
        top = newNode
    End Sub

    ' Pop the item from the top of the stack
    Public Function Pop() As String
        Dim poppedItem As String

        If IsEmpty() Then
            Console.WriteLine(vbLf & "Stack is empty - nothing to pop")
            poppedItem = Nothing
        Else
            ' Get data and change pointer to next item
            poppedItem = top.GetData()
            top = top.GetNext()
        End If

        Return poppedItem
    End Function

    ' Peek the item at the top of the stack
    Public Function Peek() As String
        Dim peekedItem As String

        If IsEmpty() Then
            Console.WriteLine(vbLf & "Stack is empty - nothing to peek")
            peekedItem = Nothing
        Else
            ' Get data from the top of the stack
            peekedItem = top.GetData()
        End If

        Return peekedItem
    End Function

    ' Output the state of the stack
    Public Sub OutputStack()
        ' Testing
        If top IsNot Nothing Then
            Console.WriteLine("------ State of the stack ------")
            Dim currentNode = top
            currentNode.OutputNode()
            While currentNode.GetNext() IsNot Nothing
                currentNode = currentNode.GetNext()
                currentNode.OutputNode()
            End While
            Console.WriteLine()
        End If
    End Sub
End Class

Module program

    ' The Main method is the entry point for all C# programs
    Sub Main()
        ' Initialise the stack
        Dim myStack As Stack = New Stack()

        ' Insert test data onto the stack
        myStack.Push("Julie")
        myStack.OutputStack()

        myStack.Push("Rey")
        myStack.OutputStack()

        ' Peek the top item of the stack
        Dim peekedItem As String = myStack.Peek()
        Console.WriteLine($"Peeked top of the stack: {peekedItem}")
        myStack.OutputStack()

        ' Insert more test data onto the stack
        myStack.Push("Habib")
        myStack.OutputStack()

        myStack.Push("Sabrina")
        myStack.OutputStack()

        ' Pop the top item from the stack
        Dim poppedItem As String = myStack.Pop()
        Console.WriteLine($"Popped {poppedItem}")
        myStack.OutputStack()

        ' Insert more test data onto the stack
        myStack.Push("Mina")
        myStack.OutputStack()

        myStack.Push("Eirini")
        myStack.OutputStack()
    End Sub


End Module

File: data-structures/stacks/java/stack_linked_list_OOP.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

// A class for a node in a linked list
class Node
{
    private String data;
    private Node next;

    // Constructor method
    public Node(String itemData)
    {
        data = itemData;
    }

    public String getData()
    {
        return data;
    }

    public Node getNext()
    {
        return next;
    }

    public void setNext(Node newNext)
    {
        next = newNext;
    }

    // Testing
    public void outputNode()
    {
        String nextData = "null";
        if (next != null) {
            nextData = next.data;
        }
        System.out.println("Data: " + data + ", next: " + nextData);
    }
}


// A stack implemented as a linked list
class Stack
{
    private Node top;

    // Check if the stack is empty
    public boolean isEmpty()
    {
        if (top == null)
            return true;
        else
            return false;
    }

    // Push a new item onto the stack
    public void push(String data)
    {
        Node newNode = new Node(data);

        // Check if the stack is NOT empty
        if (!isEmpty()) {
            // Adjust the pointer to the next node
            newNode.setNext(top);
        }

        // Set top to point to the new node
        top = newNode;
    }

    // Pop the item from the top of the stack
    public String pop()
    {
        String poppedItem;

        if (isEmpty()) {
            System.out.println("\nStack is empty - nothing to pop");
            poppedItem = "";
        }
        else {
            // Get data and change pointer to next item
            poppedItem = top.getData();
            top = top.getNext();
        }

        return poppedItem;
    }

    // Peek the item at the top of the stack
    public String peek()
    {
        String peekedItem;

        if (isEmpty()) {
            System.out.println("\nStack is empty - nothing to peek");
            peekedItem = "";
        }
        else {
            // Get data from the top of the stack
            peekedItem = top.getData();
        }

        return peekedItem;
    }

    // Output the state of the stack
    public void outputStack()
    {
        // Testing
        if (top != null) {
            System.out.println("------ State of the stack ------");
            Node currentNode = top;
            currentNode.outputNode();
            while (currentNode.getNext() != null) {
                currentNode = currentNode.getNext();
                currentNode.outputNode();
            }
            System.out.println();
        }
    }
}


class StacksExample
{
    // The main method is the entry point for all Java programs
    public static void main(String[] args)
    {
        System.out.println("### Stack (OOP) ###\n");

        // Initialise the stack
        Stack myStack = new Stack();

        // Insert test data onto the stack
        myStack.push("Julie");
        myStack.outputStack();

        myStack.push("Rey");
        myStack.outputStack();

        // Peek the top item of the stack
        String peekedItem = myStack.peek();
        System.out.println("Peeked top of the stack: " + peekedItem);
        myStack.outputStack();

        // Insert more test data onto the stack
        myStack.push("Habib");
        myStack.outputStack();

        myStack.push("Sabrina");
        myStack.outputStack();

        // Pop the top item from the stack
        String poppedItem = myStack.pop();
        System.out.println("Popped " + poppedItem);
        myStack.outputStack();

        // Insert more test data onto the stack
        myStack.push("Mina");
        myStack.outputStack();

        myStack.push("Eirini");
        myStack.outputStack();
    }


}